packages/checkout/blocks-registry/index.ts

```diff
@@ -0,0 +1,194 @@
+		);
+	}
+};
+
+/**
+ * Validation to ensure an area exists.
+ */
+const assertValidArea = ( area: string ): void => {
+	if ( ! registeredBlocks.hasOwnProperty( area ) ) {
+		throw new Error(
+			`Incorrect value for the "area" argument. It was a ${ area }, but must be one of ${ Object.keys(
+				registeredBlocks
+			).join( ', ' ) }.`
+		);
+	}
+};
+
+/**
+ * Validate the block name.
+ *
+ * @throws Will throw an error if the blockname is invalid.
+ */
+const assertBlockName = ( blockName: string ): void => {
+	assertType( 'blockName', blockName, 'string' );
+
+	if ( ! blockName ) {
+		throw new Error(
+			`Value for the blockName argument must not be empty.`
+		);
+	}
+};
+
+/**
+ * Asserts that an option is of the given type. Otherwise, throws an error.
+ *
+ * @throws Will throw an error if the type of the option doesn't match the expected type.
+ */
+const assertOption = (
+	options: Record< string, unknown >,
+	optionName: string,
+	expectedType: string
+): void => {
+	const actualType = typeof options[ optionName ];
+
+	if ( expectedType === 'array' ) {
+		if ( ! Array.isArray( options[ optionName ] ) ) {
+			throw new Error(
+				`Incorrect value for the ${ optionName } argument when registering a checkout block component. It was a ${ actualType }, but must be an array.`
+			);
+		}
+	} else if ( actualType !== expectedType ) {
+		throw new Error(
+			`Incorrect value for the ${ optionName } argument when registering a checkout block component. It was a ${ actualType }, but must be a ${ expectedType }.`
+		);
+	}
+};
+
+/**
+ * Asserts that an option is a valid react element or lazy callback. Otherwise, throws an error.
+ *
+ * @throws Will throw an error if the type of the option doesn't match the expected type.
+ */
+const assertBlockComponent = (
+	options: Record< string, unknown >,
+	optionName: string
+) => {
+	const optionValue = options[ optionName ];
+
+	if ( optionValue ) {
+		if ( typeof optionValue === 'function' ) {
+			return;
+		}
+		if (
+			isObject( optionValue ) &&
+			optionValue.$$typeof &&
+			optionValue.$$typeof === Symbol.for( 'react.lazy' )
+		) {
+			return;
+		}
+	}
+	throw new Error(
+		`Incorrect value for the ${ optionName } argument when registering a block component. Component must be a valid React Element or Lazy callback.`
+	);
+};
+
+/**
+ * Adds a block (block name) to an area, if the area exists. If the area does not exist, an error is thrown.
+ */
+const registerBlockForArea = (
+	area: keyof RegisteredBlocks,
+	blockName: string
+): void | Error => {
+	assertValidArea( area );
+	registeredBlocks = {
+		...registeredBlocks,
+		[ area ]: [ ...registeredBlocks[ area ], blockName ],
+	};
+};
+
+/**
+ * Get a list of blocks available within a specific area.
+ */
+export const getRegisteredBlocks = (
+	area: keyof RegisteredBlocks
+): Array< string > => {
+	assertValidArea( area );
+	return registeredBlocks[ area ];
+};
```

Comment:
It appears that this method is exposed publicly (via the checkout package). Do we have some concern that this then makes the array of registered blocks available for mutation by third parties (which could lead to unexpected results). I wonder if this is something we should protect against by either cloning or freezing the returned value?

Reply:
I don't understand this point. This is returning a value from `registeredBlocks` which is not exported. The only way to edit this is to go through `registerCheckoutBlock` which is exported as an extensibility API. Can you explain further?

Reply:
Yup. `packages/checkout/index.js` [exports everything from `packages/checkout/blocks-registry/index.ts`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4268/files#diff-f9bb35272358eefbb66edc83a6f637c19796d51e97ff3bf96f6112cea2f6ef2bR6). Thus `getRegisteredBlocks` is exposed on whatever global is attached to the `packages/checkout` build. `getRegisteredBlocks` will return a mutable array for the provided area thus exposing the internals of the registered blocks that can be mutated.

So if I'm understanding correctly, consumer code could do something like this (mutating any registered blocks):

```js
const registeredBlockFields = wc.blocksCheckout.getRegisteredBlocks( 'fields' );
registeredBlockFields[] = 'some/other-block';

// or
delete( registeredBlockFields[0] );
```

Reply:
Ok. I couldn't repro but I spreaded to be safe in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4268/commits/2bfe6deda4ef3f94972466290b3d7484e13c3182